locator/tablets.hh

```diff
@@ -0,0 +1,231 @@
+
+using token = dht::token;
+
+// Identifies tablet within the scope of a single tablet_map,
+// which has a scope of (table_id, token metadata version).
+// Different tablets of different tables can have the same tablet_id.
+// Different tablets in subsequent token metadata version can have the same tablet_id.
+// When splitting a tablet, one of the new tablets (in the new token metadata version)
+// will have the same tablet_id as the old one.
+enum class tablet_id : size_t;
+
+struct tablet_replica {
+    host_id host;
+    shard_id shard;
+
+    bool operator==(const tablet_replica&) const = default;
+};
+
+std::ostream& operator<<(std::ostream&, const tablet_replica&);
+
+using tablet_replica_set = utils::small_vector<tablet_replica, 3>;
+
+/// Stores information about a single tablet.
+struct tablet_info {
+    tablet_replica_set replicas;
+
+    std::optional<shard_id> get_shard(host_id host) const {
+        for (auto&& r : replicas) {
+            if (r.host == host) {
+                return r.shard;
+            }
+        }
+        return std::nullopt;
+    }
+
+    bool operator==(const tablet_info&) const = default;
+};
+
+/// Used for storing tablet state transition during topology changes.
+/// Describes transition of a single tablet.
+struct tablet_transition_info {
+    tablet_replica_set next;
+    tablet_replica pending_replica; // Optimization (next - tablet_info::replicas)
+
+    bool operator==(const tablet_transition_info&) const = default;
+};
+
+/// Stores information about tablets of a single table.
+///
+/// The map contains a constant number of tablets, tablet_count().
+/// Each tablet has an associated tablet_info, and an optional tablet_transition_info.
+/// Any given token is owned by exactly one tablet in this map.
+///
+/// A tablet map describes the whole ring, it cannot contain a partial mapping.
+/// This means that the following sequence is always valid:
+///
+///    tablet_map& tmap = ...;
+///    dht::token t = ...;
+///    tablet_id id = tmap.get_tablet_id(t);
+///    tablet_info& info = tmap.get_tablet_info(id);
+///
+/// A tablet_id obtained from an instance of tablet_map is valid for that instance only.
+class tablet_map {
+public:
+    using tablet_container = utils::chunked_vector<tablet_info>;
+private:
+    // The implementation assumes that _tablets.size() is a power of 2:
+    //
+    //   _tablets.size() == 1 << _log2_tablets
+    //
+    tablet_container _tablets;
+    size_t _log2_tablets; // log_2(_tablets.size())
+    std::unordered_map<tablet_id, tablet_transition_info> _transitions;
+public:
+    /// Default-constructed tablet_map is invalid and should not be accessed to obtain information.
+    /// Provided for so that it can be used in std::unordered_map<>.
+    tablet_map();
+
+    /// Constructs a tablet map.
+    ///
+    /// \param tablet_count The desired tablets to allocate. Must be a power of two.
+    explicit tablet_map(size_t tablet_count);
```

Comment:
For some users it would, for some it would be the opposite (we would push the check and transformation to them). The persistence layer reads the count from the system table. I don't want to store log2 size there because that's an ABI and it would limit us in the future when we relax this constraint.